### ISSUE TITLE
Accept regex literals in client languages

### ISSUE THREAD

**Issue by mlucy** [rethinkdb/rethinkdb#4308]
See https://github.com/rethinkdb/rethinkdb/pull/4305 .

If we do this we should probably do it for all three client languages.

Anyone have any thoughts on this?  It probably makes sense to most people to be able to pass regular expressions to `match`, but on the other hand the syntax RethinkDB accepts is different in some ways from the regex syntax in the clients, and I could see that confusing people (especially if they use the same regex for local matching and it returns different results).

**Comment by timmaxw** [comment]
How different is RethinkDB's regex syntax from the client regex syntaxes?

**Comment by deontologician** [comment]
we use re2, which doesn't support backreferences. JavaScript and I'm pretty sure at least Python do support them.

**Comment by marshall007** [comment]
Yea I think there are a couple escape sequences that differ slightly as well. ex. `\uFFFF` in JS would be `\x{FFFF}` in RE2.

The primary motivation behind this was to abstract users away from the `(?<flags>)` syntax RE2 uses and allow them to use the native syntax they're accustomed to instead.

> If we do this we should probably do it for all three client languages.

Agreed, however I'm not sure this would add much value in the Python driver since Python doesn't really have a concept of regex literals and flags are already specified using the `(?<flags>)` syntax. I guess we could still support passing compiled regexes for consistency.

**Comment by danielmewes** [comment]
Thanks for the idea and for submitting a pull request for this @marshall007 .

I'm quite sceptical about the interface because of the different syntax though. If native regular expressions are accepted, I think users will assume that `r.match` also interprets them in the same way as the native environment does. And it sounds like that might not always be the case.

If we can make reasonably sure that we will only ever *reject* an otherwise valid regexp but never yield a *different* result I would be fine with this though.

**Comment by marshall007** [comment]
> If we can make reasonably sure that we will only ever reject an otherwise valid regexp but never yield a different result I would be fine with this though.

@danielmewes so you're saying when the user passes in a native regex containing syntax that is unsupported in RE2 we somehow detect that and throw? That sounds tricky to implement on the client without just hard-coding in a bunch of explicit checks for known conflicts. Did you have any thoughts on how we might go about this?

**Comment by marshall007** [comment]
I just stumbled across [`node-re2`](https://github.com/uhop/node-re2) which allows you to construct RE2 expressions by passing in a native `RegExp`. If you look at his [implementation (link)](https://github.com/uhop/node-re2/blob/master/lib/new.cc#L26-L91) it looks like the only cases he handles are translating `\u` and `\c` to `\x`. Assuming these are the only cases we need to handle and RE2 on the server will just bomb out on any unsupported syntax (ex. backreferences as @deontologician  pointed out) I think we can handle these two translations transparently and call it a day.

**Comment by danielmewes** [comment]
@marshall007 
> @danielmewes so you're saying when the user passes in a native regex containing syntax that is unsupported in RE2 we somehow detect that and throw? That sounds tricky to implement on the client without just hard-coding in a bunch of explicit checks for known conflicts. Did you have any thoughts on how we might go about this?

I don't have any great ideas, but mostly because I don't know much about the differences between RE2 and standard JS regexes.
Assuming `node-re2` is reasonably complete and those are the only types of rewriting that we need to do, that doesn't sound too bad.

**Comment by mlucy** [comment]
I think the differences are much more substantial than that.  Javascript regexes support backreferences, don't they?  There's no way to rewrite those to work with RE2.

**Comment by danielmewes** [comment]
@mlucy The idea is to distinguish two different types of differences:
1. cases where a given regular expression is valid in both JS and RE2 syntax, but has different semantics
2. cases where a given regular expression is valid in JS but not in RE2

We would need to translate and/or reject differences of type 1 explicitly in the driver, but it would be ok to not do anything about differences of type 2 I think. Sending such a regex would simply result in an error, which seems ok. I think backreferences are of type 2.

**Comment by marshall007** [comment]
I've updated my PR based on our discussion. I added support for translating `\u` and `\c` escape sequences to the `\x{...}` supported by RE2. It turns out that both of these are actually type (2):

- `\u`: has different meaning but is currently unsupported
- `\c`: has same meaning but is currently unsupported

So given our rule of thumb we don't really *need* either of these rewrites, but I would say we should at least keep the `\u` since it's fairly common and would future-proof ourselves against RE2 updates (since it would have different meaning if supported). I don't have strong opinions on keeping the `\c` one though.

I've also confirmed that backtraces are definitely type (2), just note that error messages in type (2) cases might be a little hit-or-miss:

```js
r.expr("hello world").match(/((hello) (world)) \2/)
// SyntaxError: Octal literals are not allowed in strict mode.
```

**Comment by marshall007** [comment]
**tl;dr:** I wasn't able to find any type (1) cases that require translation but I think we should still translate `\u` since it's fairly common and if we don't support it there wouldn't be a portable way to match unicode characters.

**Comment by danielmewes** [comment]
Cool, thanks for doing the research and updating your implementation @marshall007 .

I'd like to give this a bit more time for discussion after we've shipped the upcoming 2.1 release, though I personally think that considering your findings we should probably support this at least for JavaScript.

**Comment by deontologician** [comment]
Maybe we could try to detect backreferences and throw a nicer error. It looks like it will be just a matter of rejecting anything with `/\\\d+` in the regex string.

**Comment by marshall007** [comment]
@deontologician I didn't realize this before, but that particular error is actually being thrown on the client before the query executes. It has something to do with `strict` mode in JS. I wasn't able to reproduce it by pulling out the relevant code, so I'm not really sure what's going on here.

```js
(function() {
  'use strict';
  
  var translateRegex = function(pattern) {
    var flags, source;
    flags = pattern.toString().match(/[im]*$/)[0];
    source = pattern.source.replace(/\\u([0-9A-F]{4})/gi, '\\x{$1}').replace(/\\c([A-Z])/g, function(match, char) {
      var hex;
      hex = (char.charCodeAt(0) - 64).toString(16);
      return "\\x{" + hex + "}";
    });
    if (flags) {
      return "(?" + flags + ")" + source;
    } else {
      return source;
    }
  };
  
  return translateRegex(/((hello) (world)) \2/);
}());
```

**Comment by marshall007** [comment]
Ok, so it turns out that backreferences are actually a type (1) case. When I first saw `SyntaxError: Octal literals are not allowed in strict mode.` I assumed it was RE2 complaining about unsupported syntax, but as mentioned above, it's really coming from JS.

What's interesting is that you'd only ever encounter this exception when running in the data explorer due to how the query gets `eval`'d.

This means we would definitely need to detect backreferences and throw an error like @deontologician suggested. Currently, (and assuming you're not using the data explorer) backreferences get passed through and interpreted as octal literal by RE2.

**Comment by deontologician** [comment]
Oof, nice catch.

**Comment by marshall007** [comment]
Just updated my PR to handle this. When backreferences are detected it now throws:

```
RqlDriverError: Cannot convert RegExp '/a(,)b\1/' to RQL: backreferences are not supported.
```

**Comment by mlucy** [comment]
So, how sure are we that there aren't going to be subtle differences between what regexps meant in the client language and what they mean in ReQL?  Honestly I'm still a little spooked by the idea.

**Comment by deontologician** [comment]
It's kinda complicated. I looked for a good test-suite of javascript regex syntax, but couldn't find one. There were some for python and re2 though. Ideally we'd run through a ton of existing syntax examples on test strings, validate that either they both matched the same string (type 0), the re2 library threw an error (type 1) or the conversion function threw an error (type 2).

That's a lot of work, and even in the languages with regex literals, it's saving a few characters at most. I think this is neat, but maybe not worth the headache.

**Comment by marshall007** [comment]
So the original idea with this PR was to simply accept RE2-compatible expressions written using native literals (i.e. it maps 1:1 with what you would have passed `r.match` as a string). I think this is easier to maintain, understand,  and simpler to explain than trying to patch up the differing syntax. I'd be in favor of reverting the translation logic and just adding a clarification in the docs:

> JavaScript and RE2 regular expressions have slightly different syntax and a different set of available flags. A valid `RegExp` interpreted by the server may not compile, or may match a different set of strings in JavaScript than when used in a RethinkDB query.

**Comment by marshall007** [comment]
I looked around for some prior-art and this is how it's handled in all the MongoDB client drivers I looked at (Node, C#, Java, and Python). MongoDB uses PCRE on the server and none of the clients did any translations when serializing their native regular expression objects in queries.

http://docs.mongodb.org/manual/reference/operator/query/regex/

**Comment by danielmewes** [comment]
Does this add any additional convenience apart from the slightly different style of writing the regex string and not having to escape `\`?

**Comment by marshall007** [comment]
@danielmewes nope, that's the only purpose.

**Comment by danielmewes** [comment]
Thanks @marshall007. It's really unfortunate to discard your pull request, but I think given that and some of the difficulties with the syntax differences, I'd prefer not to accept the JavaScript Regexp literals in `match`.
Especially new users sometimes get confused about which operations are executed on the server and which on the client in my experience. Accepting the native Regexp syntax but with slightly different meanings might add to that confusion.

**Comment by marshall007** [comment]
@danielmewes no worries, thanks to everyone for the discussion!

**Comment by danielmewes** [comment]
We're not going to support this for the time being. Thanks everyone (especially @marshall007 ) for the work on this!
